### ISSUE TITLE
Loss mismatch related issues

### ISSUE THREAD

**Issue by Jiyu001-zhang** [AlonzoLeeeooo/LCDG#4]
In the paper, it is written that the loss is DM + cond, but the code only shows the cond  related loss. Looking forward to hearing from you!
![Snipaste_2024-07-09_11-35-51](https://github.com/AlonzoLeeeooo/LCDG/assets/57984638/ae22b046-aacd-439d-96fe-21257e76ebca)
![Snipaste_2024-07-09_11-36-28](https://github.com/AlonzoLeeeooo/LCDG/assets/57984638/c6b837f5-f45f-4c65-b904-de0ae6025a5c)

**Comment by AlonzoLeeeooo** [comment]
Hi @Jiyu001-zhang ,

Thank you for your attention of our work!

In fact, we only use the condition loss (i.e., MSE loss between reconstructed and ground truth condition) to optimize the condition adpater, corresponding to Eq. (4) in our main paper.
![image](https://github.com/AlonzoLeeeooo/LCDG/assets/68578066/fc0e03a8-d7db-485f-a08a-f5593147e75e)


As is pointed out in Fig. 3, the equation to compute $\widehat{\epsilon}_{t}$, i.e., Eq (6) in our main paper, refers to the conditional score adaptation process. Please refer to the code of modified DDIM sampler, i.e., line 251 to line 276 in [`ldm/models/diffusion/ddim.py`](https://github.com/AlonzoLeeeooo/LCDG/blob/main/ldm/models/diffusion/ddim.py).
```python
# TODO: compute conditioning & re-compute classifier-free guidance
          if cond_fn is not None:
              # TODO: only add condition within the truncation steps, in order to manipulate the geometry of samples
              if t > torch.tensor(truncation_steps).cuda(t.device):
                  # TODO: add condition only in the first T steps
                  if return_pred_cond:
                        gradient, pred_edge = cond_fn(x=x_in.chunk(2)[0], c=c_in.chunk(2)[1], t=t_in.chunk(2)[0], cond_configs=cond_configs, blocks_indexes=blocks_indexes,
                                                    target_cond=target_cond, model=cond_model, diffusion_model=self.model, return_pred_cond=return_pred_cond,
                                                    criterion=cond_criterion)
                    else:
                        gradient = cond_fn(x=x_in.chunk(2)[0], c=c_in.chunk(2)[1], t=t_in.chunk(2)[0], cond_configs=cond_configs, blocks_indexes=blocks_indexes,
                                                target_cond=target_cond, model=cond_model, diffusion_model=self.model, return_pred_cond=return_pred_cond,
                                                criterion=cond_criterion)
                    # if t <= torch.tensor(1000).cuda(t.device) and t >= torch.tensor(750).cuda(t.device):
                    #     print(t)
                    #     cond_scale = 5.0
                    
                    gradient = gradient * cond_scale * torch.norm(input=e_t - x_in.chunk(2)[0], p=2) / torch.norm(input=gradient, p=2)                        # TODO: normalization weight: important to keep the values balance
                    if add_cond_score:
                        e_t = e_t_uncond + unconditional_guidance_scale * (e_t - e_t_uncond) + gradient
                    else:
                        e_t = e_t_uncond + unconditional_guidance_scale * (e_t - e_t_uncond) - gradient
                else:
                    e_t = e_t_uncond + unconditional_guidance_scale * (e_t - e_t_uncond)
            else:
                e_t = e_t_uncond + unconditional_guidance_scale * (e_t - e_t_uncond)
```

In the above code, ` e_t = e_t_uncond + unconditional_guidance_scale * (e_t - e_t_uncond) + gradient` gives a direct illustration of Eq. (6), combining with classifier-free guidance.

Hope this explanation addresses your concern! Please feel free to contact us if you have any further questions.

Best regards

**Comment by AlonzoLeeeooo** [comment]
Hi @Jiyu001-zhang ,

This issue will be closed for now, since the problem is explained above. We always welcome you to leave further messages in this issue if you have other related questions about our work.

Best
